File: riptide_hardware/riptide_hardware2/depth_converter.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
#include "riptide_msgs2/msg/depth.hpp"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "Eigen/Dense"

using std::placeholders::_1;

class DepthConverter : public rclcpp::Node {
public:
    DepthConverter() : Node("depth_converter") {
        auto qos = rclcpp::SensorDataQoS();
        sub_ = this->create_subscription<riptide_msgs2::msg::Depth>(
            "state/depth/raw", qos, std::bind(&DepthConverter::depth_callback, this, _1));
        pub_ = this->create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>(
            "depth/pose", qos);
        tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

        std::string full_namespace = this->get_namespace();
        namespace_ = full_namespace.substr(1); // Skip the leading slash

    }

private:
    void depth_callback(const riptide_msgs2::msg::Depth::SharedPtr msg) {
        try {
            auto b2o_transform = tf_buffer_->lookupTransform("odom", namespace_ + "/base_link", tf2::TimePointZero);
            Eigen::Quaterniond b2o_orientation(
                b2o_transform.transform.rotation.w,
                b2o_transform.transform.rotation.x,
                b2o_transform.transform.rotation.y,
                b2o_transform.transform.rotation.z);
            Eigen::Matrix3d b2o_matrix = b2o_orientation.toRotationMatrix();

            if (!b2p_vector_.has_value()) {
                auto pressure_transform = tf_buffer_->lookupTransform(namespace_ + "/pressure_link", namespace_ + "/base_link", tf2::TimePointZero);
                b2p_vector_ = Eigen::Vector3d(
                    pressure_transform.transform.translation.x,
                    pressure_transform.transform.translation.y,
                    pressure_transform.transform.translation.z);
            }

            double added_depth = b2o_matrix.col(2).dot(b2p_vector_.value());

            auto out_msg = geometry_msgs::msg::PoseWithCovarianceStamped();
            out_msg.header = msg->header;
            out_msg.header.frame_id = "odom";
            out_msg.pose.pose.position.z = msg->depth + added_depth;
            out_msg.pose.covariance[14] = msg->variance;
            out_msg.header.stamp = this->get_clock()->now();
            pub_->publish(out_msg);
        } catch (const tf2::TransformException& ex) {
            RCLCPP_WARN(this->get_logger(), "TF2 error: %s", ex.what());
        }
    }

    rclcpp::Subscription<riptide_msgs2::msg::Depth>::SharedPtr sub_;
    rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr pub_;
    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
    std::optional<Eigen::Vector3d> b2p_vector_;
    std::string namespace_;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DepthConverter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: riptide_hardware/riptide_hardware2/depth_converter.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from rclpy.time import Time
from geometry_msgs.msg import PoseWithCovarianceStamped
from riptide_msgs2.msg import Depth
import tf2_ros
import transforms3d as tf3d
import numpy as np


class depthConverter(Node):
    def __init__(self):
        super().__init__('depth_converter')
        self.sub = self.create_subscription(
            Depth, "state/depth/raw", self.depthCb, qos_profile_sensor_data)
        self.pub = self.create_publisher(
            PoseWithCovarianceStamped, "depth/pose", qos_profile_sensor_data)
        self.namespace = self.get_namespace()[1:]
        self.tfBuffer = tf2_ros.Buffer()
        self.listener = tf2_ros.TransformListener(self.tfBuffer, self)
        self.b2pVector = None

    def depthCb(self, msg):
        try:
            # Rotation from base frame to odom
            b2oOrientation = self.tfBuffer.lookup_transform(
                'odom', self.namespace+'/base_link', Time()).transform.rotation
            b2oMatrix = tf3d.quaternions.quat2mat(
                [b2oOrientation.w, b2oOrientation.x, b2oOrientation.y, b2oOrientation.z])[:3, :3]

            if self.b2pVector is None:
                # Offset to pressure sensor
                pressureOffset = self.tfBuffer.lookup_transform(
                    self.namespace+'/pressure_link', self.namespace+'/base_link', Time()).transform.translation
                self.b2pVector = [pressureOffset.x,
                                  pressureOffset.y, pressureOffset.z]
        except (tf2_ros.LookupException, tf2_ros.ConnectivityException, tf2_ros.ExtrapolationException) as ex:
            self.get_logger().warning(str(ex))
            return

        # Rotate pressure sensor offset into odom frame and get additional depth from offset
        # TODO: Calculate uncertainty of this measure
        addedDepth = np.dot(b2oMatrix, self.b2pVector)[2]

        # Publish z offset from odom to base_link (depth)
        outMsg = PoseWithCovarianceStamped()
        outMsg.header = msg.header
        outMsg.header.frame_id = "odom"
        outMsg.pose.pose.position.z = msg.depth + addedDepth
        outMsg.pose.covariance[14] = msg.variance
        outMsg.header.stamp = self.get_clock().now().to_msg()
        self.pub.publish(outMsg)


def main(args=None):
    rclpy.init(args=args)
    node = depthConverter()
    rclpy.spin(node)


if __name__ == '__main__':
    main()
